db/migrate/20200310091539_add_last_error_to_deployments.rb

```diff
@@ -0,0 +1,5 @@
+class AddLastErrorToDeployments < ActiveRecord::Migration[5.2]
+  def change
+    add_column :deployments, :last_error, :string
```

Comment:
will this column store last error log?
can we rename `last_error_log` ?
also would it be better if we also have `last_error_at` timestamp field?

Reply:
Deployment scheduler processing will be done every 5 seconds.

If a deployment can't be processed and raised an exception, it will be taken and the exception message will be put to the `last_error` field of the deployment.

If the deployment can be processed, the `last_error` field will be cleared. So, it's not a log, but just an error message of last deployment scheduler processing.

Reply:
I'm asking for your opinion. Is that a good design?

And if we choose that design, do we need to store `last_error_at`? My opinion is we don't have to since the last deployment scheduler processing was successful, `last_error` will be cleared. But I want to know your opinion.

Reply:
usually in this situation, we can STDERR/STDOUT to console / logfile
I wonder why we cannot just do that and have to store the error message in database. This is a bit uncommon, unless I'm missing something.

Reply:
We'll log it to STDERR, STDOUT too for more verbose information. (it should be, maybe I forgot to log it)

We have a use case here. For example, someone sets a wrong function name of `$pf-meta:` script in deployment definition field, e.g.:
```
$pf-meta:unknown-function
```

Deployment scheduler will kick in and process that field, resulting in an error because there's no function named `unknown-function` in `$pf-meta:` script and an exception will be thrown.

That person may be confused about why the deployment isn't creating containers or why it's not updating the containers. In our use case, that person can see that there's `last_error` that said `unknown-function` can't be recognized as a valid function. So, that person can fix the error.

When the error has already fixed, the deployment scheduler will successfully process the field and will clear the `last_error` field, signify that the error has gone and the deployment can be processed.

What's your opinion @giosakti? Is it good enough? And should we change the field name to be `last_error_message`?

Reply:
`That person may be confused about why the deployment isn't creating containers or why it's not updating the containers`

But I thought, it will be outputted to STDOUT?

`When the error has already fixed, the deployment scheduler will successfully process the field and will clear the last_error field, signify that the error has gone and the deployment can be processed.`

Isn't it better if we have a state field instead with success/error code?

Reply:
The actual purpose is actually convenience.

So the person can just see the Pathfinder UI, acknowledging that his/her deployment is failing and he also **knows the summary of the error without looking at the log file of Pathfinder.** So he can take a **quick action** to fix his deployment definition.

As we know that deployment definition is an **user input.** The user should be able to know **the flaws of his input from the UI also.** If we just put the state, whether it's an error or not, the user **needs to check the log of Pathfinder scheduler** even the error is **because of he's putting wrong value** in deployment definition user input.

What do you think @giosakti?

Reply:
ok then since you said this is for UI then it makes sense. Let's rename it to `last_error_msg`.

Also I still think `last_error_at` is important since this is for UI then user need to know the error message is supposed to point to which error in time. Log / error message usually always comes in pair, timestamp and the message. Other option is to just embed timestamp in the body of the message itself.

I assume we don't need timestamp in error logs in console since it will automatically be added.

Reply:
Noted mas Gio. Thanks!

Reply:
We have modified this in a new migration. @giosakti